Run RSpec tests in custom engines (and some fixes, cleanup)

kristinalim opened the pull request:
#### What? Why?

Closes #3085 

RSpec tests of the engine "web" are not being run by the commands we use.

This PR includes the following changes:

1. Allow tests in "web" engine to be run separately. Using the main app's `spec_helper.rb` for now, while we are still slowly moving to Rails engines.
2. Add Rake task `openfoodnetwork:specs:engines:rspec` to run RSpec tests for all engines.
3. Add Rake tasks `openfoodnetwork:specs:engines:ENGINE_NAME:rspec` to run RSpec tests of a specific engine.

For your reference, here are Semaphore builds which already run the engine tests on my fork:

- `cleanup-test_scripts` - https://semaphoreci.com/kristinalim/openfoodnetwork/branches/cleanup-test_scripts/builds/14
- `cleanup-test_scripts-failure` which adds a failing engine test - https://semaphoreci.com/kristinalim/openfoodnetwork/branches/cleanup-test_scripts-failure/builds/5

#### What should we test?

Only for developers:

1. Introduce a failing test in `engines/web/spec/`.
2. The failure should reflect in Semaphore.

The `.travis.yml` configuration file currently [does not build successfully](https://github.com/openfoodfoundation/openfoodnetwork/pull/3082#pullrequestreview-176609809).

#### Release notes

- Allow tests in "web" engine to be run with `bundle exec rspec` from the engine path.
- Add Rake task `openfoodnetwork:specs:engines:rspec` to run RSpec tests for all engines.
- Add Rake tasks `openfoodnetwork:specs:engines:ENGINE_NAME:rspec` to run RSpec tests of a specific engine.

Changelog Category: Fixed

#### Documentation updates

1. [Continuous Integration](https://github.com/openfoodfoundation/openfoodnetwork/wiki/Continuous-Integration) - One of the jobs should be updated to run the Rake task `openfoodnetwork:specs:engines:rspec`.

#### After merge

1. Communicate this in `#dev` Slack channel.
2. Update project settings in Semaphore official account to run the Rake task `openfoodnetwork:specs:engines:rspec`.

kristinalim commented:
@sauloperez As far as I understand, the motivation behind packaging code into engines is to make them work independently. Running them in a single `rspec` process would not expose places where their libraries are still tightly coupled.

Although, I think at the start, most engines would not be able to stand on their own and could therefore still use the main application's `spec_helper.rb` which includes the main application (and all the other engines). For example, running `bundle exec rspec` in "web" raises `NoMethodError:
  undefined method `scenario' for #<Class:...>`, and it does not define nor require `BaseController` which it uses.

Do you think this is okay?

sauloperez commented:
I understand what you say. My point was just implementation. Instead of adding 3 new bash scripts I would give it a try at sticking to rake to solve what `script/specs/rspec_engines` covers.

kristinalim commented:
Thanks, I misunderstood! Yes, that sounds great. So a non `:environment` Rake task, for ruby scripting instead of bash.

sauloperez commented:
> So a non :environment Rake task, for ruby scripting instead of bash.

If they don't need to load the environment :+1:, but as you said they will not be able to stand on their own, right? so we might need it at least for now. In any case, rake supports both ways.

daniellemoorhead commented:
Hey @kristinalim 👋 

No progress on this for 22 days...I'm wondering if this needs to be closed for now and moved back to Dev Ready until your to do list is clear again?

#toomanythingsinthepipe
#toottoot

kristinalim commented:
Thanks @daniellemoorhead! It so happened that I finished the two other PRs in my In Dev set, so I worked on this again.

@sauloperez I changed this to use a Rake task instead of a script in `script/`. I also updated the issue description, so please read again if you're reviewing this.

The Code Climate check failure would be fixed with a rebase after #3223 is merged.

luisramos0 commented on lib/tasks/specs.rake:
```diff
@@ -0,0 +1,47 @@
+namespace :openfoodnetwork do
```
@sauloperez we wanted to get rid of this namespace in the rake tasks right?
it was discussed in @mkllnk's sample_data PR #3209. we should keep both tasks coherent in terms of namespaces. do you guys know what's the best approach here?

sauloperez replied:
Thanks for pointing out. I thought about it but I prefered to not annoy again with that topic. My opinion is that we should get rid of it.

luisramos0 replied:
lets get rid of it then :-D

mkllnk replied:
Let's have the discussion on https://community.openfoodnetwork.org/t/rake-task-naming/1531.
